Add push notifications

fabioscsilva opened the pull request:
This closes #242

fabioscsilva committed:
add notifications draft

diogoteix committed:
notifications modal

fabioscsilva committed:
Fix push notifications

ruisaraiva19 commented on Trackovid19-web/package.json:
```diff
@@ -26,6 +26,7 @@
     "@angular/platform-browser": "9.0.7",
     "@angular/platform-browser-dynamic": "9.0.7",
     "@angular/router": "9.0.7",
+    "@angular/service-worker": "9.0.3",
```
This package version should match the other Angular packages versions `9.0.7`

fabioscsilva replied:
👍

ruisaraiva19 commented on Trackovid19-web/src/app/app.component.ts:
```diff
@@ -25,4 +30,15 @@ export class AppComponent {
+    if (this.swUpdate.isEnabled) {
+      this.swUpdate.available.subscribe(() => {
+        if (confirm('New version available. Load New Version?')) {
+          window.location.reload();
```
To update the PWA to a new version, a simple `window.location.reload()` won't do it.

fabioscsilva replied:
From what I've seen until know it works. But probably I'm forgetting something.
Can you tell me what is need?

ruisaraiva19 replied:
The gist of it is, first we send a message to the current service worker telling to activate the new service worker and them we reload the page.

fabioscsilva replied:
I'm not following you know. 
The service work is the same, what changes is the `ngsw.json` which have the list of all files and versions. Every time we enter the page, we check the file, and if something change `this.swUpdate.available` is triggered and we just need to reload the page, since the new assets were download in the meanwhile.

ruisaraiva19 replied:
Technically there are two different service worker instances, the currently being used and the new one waiting to take place of the current one.

We only need a simple page refresh if the service worker `skipWaiting` method was already called before (this method replaces the currently running service worker with the new one), which I'm not sure with the Angular service worker configuration already does that for us on the `install` service worker event.

fabioscsilva replied:
Ok, now I'm understanding your concern and I'm following you. 
The angular service worker is doing that for us 👍 

```js
           this.scope.addEventListener('install', (event) => {
                // SW code updates are separate from application updates, so code updates are
                // almost as straightforward as restarting the SW. Because of this, it's always
                // safe to skip waiting until application tabs are closed, and activate the new
                // SW version immediately.
                event.waitUntil(this.scope.skipWaiting());
            });
```

ruisaraiva19 replied:
Nice, it looks like everything is ready to go 🚀.

fabioscsilva replied:
Nice. Thanks for your review!

ruisaraiva19 commented on Trackovid19-web/src/manifest.webmanifest:
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "Covidografia",
+    "short_name": "Covidografia",
+    "theme_color": "#ffffff",
```
This variable needs to be updated. Probably to the project orange `#fa6400`.

fabioscsilva replied:
This was given by the designer. All the app background is white. So, why you think we should update it?

ruisaraiva19 replied:
The `theme_color` is not the same as `background_color`.

fabioscsilva replied:
Yes, I know. And again we want it white. 
theme_color - we want that for instance task switcher stays white
background_color - we want that the loading screen white

ruisaraiva19 commented on Trackovid19-web/src/manifest.webmanifest:
```diff
@@ -0,0 +1,21 @@
+    "background_color": "#ffffff",
+    "display": "standalone",
+    "scope": "./",
+    "start_url": "./",
```
This URL can be updated with a query parameter for analytics porpuses. Something like:
`"start_url": "./index.html?source=a2hs",`.
Setting the `source` query parameter to `a2hs` (add to home screen) is a good PWA practice.

fabioscsilva replied:
👍

ruisaraiva19 commented on Trackovid19-web/src/manifest.webmanifest:
```diff
@@ -0,0 +1,21 @@
+    "start_url": "./",
+    "icons": [
+        {
+            "src": "assets/android-chrome-192x192.png",
```
This asset does not exist.

fabioscsilva replied:
It exist, but i forgot to update the path.
Good catch!

ruisaraiva19 commented on Trackovid19-web/src/manifest.webmanifest:
```diff
@@ -0,0 +1,21 @@
+            "type": "image/png"
+        },
+        {
+            "src": "assets/android-chrome-512x512.png",
```
Same here.

fabioscsilva replied:
Same as previous

ruisaraiva19 commented on Trackovid19-web/src/index.html:
```diff
@@ -9,7 +9,7 @@
-    <link rel="manifest" href="assets/img/favicon/site.webmanifest" />
+    <link rel="manifest" href="manifest.webmanifest" />
     <link rel="mask-icon" href="assets/img/favicon/safari-pinned-tab.svg" color="#5bbad5" />
     <meta name="msapplication-TileColor" content="#da532c" />
```
This meta tag content should be updated to the project orange color `#fa6400`

fabioscsilva replied:
Same as previous, but I will update to white

ruisaraiva19 replied:
The `msapplication-TileColor` in particular shouldn't be white since the Microsoft tiles text are white as well.

fabioscsilva replied:
I will revert it to the color that designer give me from the beginning

ruisaraiva19 commented on Trackovid19-web/src/index.html:
```diff
@@ -9,7 +9,7 @@
+    <link rel="manifest" href="manifest.webmanifest" />
     <link rel="mask-icon" href="assets/img/favicon/safari-pinned-tab.svg" color="#5bbad5" />
     <meta name="msapplication-TileColor" content="#da532c" />
     <meta name="theme-color" content="#ffffff" />
```
Same here.

fabioscsilva replied:
Same as previous

ruisaraiva19 commented on Trackovid19-web/src/manifest.webmanifest:
```diff
@@ -0,0 +1,21 @@
+            "type": "image/png"
+        }
+    ]
+}
```
There are a couple more properties we can add to this manifest file.

fabioscsilva replied:
Yes, I know. But can you give more detail, which ones are useful to have?

ruisaraiva19 replied:
For example, adding `"purpose": "maskable"` to each icon object will be consistent across OS versions.

fabioscsilva replied:
I revisited the manifest file and I added some more properties, including `"purpose": "maskable"`. Can you check if this is ready now?

ruisaraiva19 replied:
Yes, looks good 👌.

ruisaraiva19 reviewed (changes requested):
Looks good so far. A couple of things should be updated/improved for consistency.

I have experience with everything related to PWA technologies. Please let me know if you need any help with this 😉.

fabioscsilva committed:
Address code review.

	- Update icon paths
	- Update service worker version
	- Update manifest start_url

fabioscsilva committed:
Update manifest file and reverted colors
